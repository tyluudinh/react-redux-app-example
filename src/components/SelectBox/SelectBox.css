.custom-select {
  * {
    font-weight: lighter !important;
  }

  &.is-focused {
    .Select-control{
      box-shadow: 0 2px 4px rgba(0,0,0,0.25), 0 3px 3px rgba(0,0,0,0.22);
    }
  }

  .Select-clear {
    color: #888888;
  }

  .Select-control {
    border-radius: 0;
    font-weight: light;
    font-size: 0.9rem;
    padding: 8px;
    border: 0;

    box-shadow: 0 1px 3px rgba(0,0,0,0.09), 0 1px 2px rgba(0,0,0,0.16);
    transition: all 0.3s cubic-bezier(.25,.8,.25,1);

    .Select-placeholder {
      padding-left: 17px;
      line-height: 50px;
    }
  }
  
  .Select-value {
    line-height: 3.2rem !important;
    padding-left: 17px !important;
  }

  .Select-placeholder {
    font-size: 0.8rem;
  }

  .Select-option {
    padding-left: 17px !important;
    font-size: 0.8rem;
  }

  .Select-value-label {
    color: #3d3d3d !important;
  }

  .Select-value {
    font-size: 0.8rem;
  }

  .custom-option--default {
    width: 100%;
    padding: 10px 18px;
    display: block;
    cursor: pointer;
    &:hover {
      background-color: #f7f7f7;
    }
  }

}

.custom-select--flat {
  box-shadow: none;
  border: 0;

  &.is-focused {
    .Select-control {
      box-shadow: none;
    }
  }
  
  .Select-control {
    box-shadow: none;
  }
}

.custom-select--passive {
  cursor: default;
  .Select-arrow-zone {
    display: none;
  }

  .Select-menu-outer {
    box-shadow: 0 1px 3px rgba(0,0,0,0.12), 0 1px 2px rgba(0,0,0,0.24);  
    border-radius: 0;
    border: 0;
    padding: 12px;
  }

  &.is-focused {
    .Select-arrow-zone {
      display: table-cell;
    }
  }
}

.custom-select__arrow {
  font-size: 0.6rem;
  margin-right: -3px;
  fill: #919191;
}

.custom-select.is-disabled {
  * { color: #aaaaaa !important}
}
