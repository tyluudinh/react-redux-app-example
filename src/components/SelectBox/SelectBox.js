import React, { Component } from "react";
import Select, { Creatable } from "react-select";
import classNames from "classnames";
import PropTypes from "prop-types";
import DownArrow from "react-icons/lib/fa/chevron-down";

import "react-select/dist/react-select.css";
import "./SelectBox.css";

class CustomValue extends Component {
  render() {
    const { color, label } = this.props.value;
    if (color)
      return (
        <span className="custom-value--default" style={{ color }}>
          {label}
        </span>
      );
    return <span>{label}</span>;
  }
}

class CustomOption extends Component {
  render() {
    const { onSelect, isFocused } = this.props;
    const { color, label, value } = this.props.option;
    if (color) {
      return (
        <span
          className={classNames("custom-option--default", {
            "is-focused": isFocused
          })}
          style={{ color }}
          onClick={e => onSelect({ value }, e)}
        >
          {label}
        </span>
      );
    }
    return <span>{label}</span>;
  }
}

export default class SelectBox extends Component {
  constructor(props) {
    super(props);
    const { allowCreate } = props;

    this.getProperSelect({ allowCreate });

    this.onChange = this.onChange.bind(this);
    this.onNewOption = this.onNewOption.bind(this);
    this._renderArrow = this._renderArrow.bind(this);
  }

  SelectComponent = Select;

  // This function gonna check whether to use Select or Creatable as SelectBox Comonent
  getProperSelect({ allowCreate }) {
    if (allowCreate) this.SelectComponent = Creatable;
  }

  onNewOption(option) {
    const { type, onNewOption } = this.props;
    return type ? onNewOption(type, option) : onNewOption(option);
  }

  onChange(option) {
    const { onChange, name } = this.props;
    const goodOption = (option === null) ? { value: null, label: '', serverSync: true } : option;
    if (goodOption && onChange) {
      return name ? onChange(name, goodOption.value) : onChange(goodOption.value);
    }
  }

  _renderArrow() {
    if (this.props.noArrow) return null;
    return (
      <span>
        <DownArrow className="custom-select__arrow down" />
      </span>
    );
  }

  render() {
    const {
      options,
      value,
      flat,
      passive,
      customItems,
      name,
      onNewOption,
      placeholder,
      disabled,
    } = this.props;
    let additionalOptions = {};
    if (customItems) {
      additionalOptions.valueComponent = CustomValue;
      additionalOptions.optionComponent = CustomOption;
    }
    if (onNewOption) {
      additionalOptions.onNewOptionClick = this.onNewOption;
    }

    return (
      <this.SelectComponent
        className={classNames(
          "custom-select",
          { "custom-select--flat": flat },
          { "custom-select--passive": passive }
        )}
        name={name}
        value={value}
        options={options}
        onChange={this.onChange}
        arrowRenderer={this._renderArrow}
        placeholder={placeholder}
        clearable={onNewOption ? true : false}
        disabled={disabled}
        {...additionalOptions}
      />
    );
  }
}

SelectBox.propTypes = {
  options: PropTypes.array.isRequired,
  value: PropTypes.string,
  onChange: PropTypes.func,
  name: PropTypes.string,
  placeholder: PropTypes.string,
  passive: PropTypes.bool,
  flat: PropTypes.bool,
  customItems: PropTypes.bool,
  allowCreate: PropTypes.bool
};
