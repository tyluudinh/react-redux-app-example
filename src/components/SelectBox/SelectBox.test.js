import React from 'react';
import Select from 'react-select';
import SelectBox from './SelectBox';

describe('Component SelectBox', () => {
  const normalOptions = [
    { value: 'one', label: 'option one' },
    { value: 'two', label: 'option two' },
    { value: 'three', label: 'option three' },
    { value: 'four', label: 'option four' },
    { value: 'all', label: 'all options' },
  ]

  test('To be rendered correctly', () => {
    // Normal selectBox
    const component = mount(<SelectBox value="one" options={normalOptions} />);
    const componentJson = toJson(component);
    expect(componentJson).toMatchSnapshot();
  })

  describe('Event Handler', () => {
    const onChangeMock = jest.fn();
    const component = shallow(<SelectBox options={normalOptions} onChange={onChangeMock} />);
    component.find(Select).simulate('change', 'selected option');
  
    test('Is called once', () => {
      expect(onChangeMock.mock.calls.length).toBe(1);
    })
    
    test('Correct value passed', () => {
      expect(onChangeMock.mock.calls[0][0]).toBe('selected option');
    })
  })
})