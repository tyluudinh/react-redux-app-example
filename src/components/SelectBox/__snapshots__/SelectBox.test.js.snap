// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component SelectBox To be rendered correctly 1`] = `
<SelectBox
  options={
    Array [
      Object {
        "label": "option one",
        "value": "one",
      },
      Object {
        "label": "option two",
        "value": "two",
      },
      Object {
        "label": "option three",
        "value": "three",
      },
      Object {
        "label": "option four",
        "value": "four",
      },
      Object {
        "label": "all options",
        "value": "all",
      },
    ]
  }
  value="one"
>
  <Select
    addLabelText="Add \\"{label}\\"?"
    arrowRenderer={[Function]}
    autosize={true}
    backspaceRemoves={true}
    backspaceToRemoveMessage="Press backspace to remove {label}"
    className="custom-select"
    clearAllText="Clear all"
    clearRenderer={[Function]}
    clearValueText="Clear value"
    clearable={false}
    closeOnSelect={true}
    deleteRemoves={true}
    delimiter=","
    disabled={false}
    escapeClearsValue={true}
    filterOptions={[Function]}
    ignoreAccents={true}
    ignoreCase={true}
    inputProps={Object {}}
    isLoading={false}
    joinValues={false}
    labelKey="label"
    matchPos="any"
    matchProp="any"
    menuBuffer={0}
    menuRenderer={[Function]}
    multi={false}
    name="form-field-name"
    noResultsText="No results found"
    onBlurResetsInput={true}
    onChange={undefined}
    onCloseResetsInput={true}
    onSelectResetsInput={true}
    openOnClick={true}
    optionComponent={[Function]}
    options={
      Array [
        Object {
          "label": "option one",
          "value": "one",
        },
        Object {
          "label": "option two",
          "value": "two",
        },
        Object {
          "label": "option three",
          "value": "three",
        },
        Object {
          "label": "option four",
          "value": "four",
        },
        Object {
          "label": "all options",
          "value": "all",
        },
      ]
    }
    pageSize={5}
    placeholder="Select..."
    required={false}
    scrollMenuIntoView={true}
    searchable={true}
    simpleValue={false}
    tabSelectsValue={true}
    value="one"
    valueComponent={[Function]}
    valueKey="value"
  >
    <div
      className="Select custom-select Select--single is-searchable has-value"
      style={undefined}
    >
      <input
        disabled={false}
        key="hidden.0"
        name="form-field-name"
        type="hidden"
        value="one"
      />
      <div
        className="Select-control"
        onKeyDown={[Function]}
        onMouseDown={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        style={undefined}
      >
        <span
          className="Select-multi-value-wrapper"
          id="react-select-2--value"
        >
          <Value
            disabled={false}
            id="react-select-2--value-item"
            instancePrefix="react-select-2-"
            onClick={null}
            value={
              Object {
                "label": "option one",
                "value": "one",
              }
            }
          >
            <div
              className="Select-value"
              style={undefined}
              title={undefined}
            >
              <span
                aria-selected="true"
                className="Select-value-label"
                id="react-select-2--value-item"
                role="option"
              >
                option one
              </span>
            </div>
          </Value>
          <AutosizeInput
            aria-activedescendant="react-select-2--value"
            aria-describedby={undefined}
            aria-expanded="false"
            aria-haspopup="false"
            aria-label={undefined}
            aria-labelledby={undefined}
            aria-owns=""
            className="Select-input"
            minWidth="5"
            onBlur={[Function]}
            onChange={[Function]}
            onFocus={[Function]}
            required={false}
            role="combobox"
            tabIndex={undefined}
            value=""
          >
            <div
              className="Select-input"
              style={
                Object {
                  "display": "inline-block",
                }
              }
            >
              <style
                dangerouslySetInnerHTML={
                  Object {
                    "__html": "input#undefined::-ms-clear {display: none;}",
                  }
                }
              />
              <input
                aria-activedescendant="react-select-2--value"
                aria-describedby={undefined}
                aria-expanded="false"
                aria-haspopup="false"
                aria-label={undefined}
                aria-labelledby={undefined}
                aria-owns=""
                className={undefined}
                id={undefined}
                onBlur={[Function]}
                onChange={[Function]}
                onFocus={[Function]}
                required={false}
                role="combobox"
                style={
                  Object {
                    "boxSizing": "content-box",
                    "width": "5px",
                  }
                }
                tabIndex={undefined}
                value=""
              />
              <div
                style={
                  Object {
                    "height": 0,
                    "left": 0,
                    "overflow": "scroll",
                    "position": "absolute",
                    "top": 0,
                    "visibility": "hidden",
                    "whiteSpace": "pre",
                  }
                }
              />
            </div>
          </AutosizeInput>
        </span>
        <span
          className="Select-arrow-zone"
          onMouseDown={[Function]}
        >
          <span>
            <FaChevronDown
              className="custom-select__arrow down"
            >
              <IconBase
                className="custom-select__arrow down"
                viewBox="0 0 40 40"
              >
                <svg
                  className="custom-select__arrow down"
                  fill="currentColor"
                  height="1em"
                  preserveAspectRatio="xMidYMid meet"
                  style={
                    Object {
                      "color": undefined,
                      "verticalAlign": "middle",
                    }
                  }
                  viewBox="0 0 40 40"
                  width="1em"
                >
                  <g>
                    <path
                      d="m37.6 18l-16.6 16.6q-0.4 0.4-1 0.4t-1-0.4l-16.6-16.6q-0.4-0.4-0.4-1t0.4-1l3.7-3.7q0.5-0.4 1-0.4t1 0.4l11.9 11.9 11.9-11.9q0.4-0.4 1-0.4t1 0.4l3.7 3.7q0.4 0.4 0.4 1t-0.4 1z"
                    />
                  </g>
                </svg>
              </IconBase>
            </FaChevronDown>
          </span>
        </span>
      </div>
    </div>
  </Select>
</SelectBox>
`;
